### ISSUE TITLE
Storybook 5 Decorator Syntax Broken

### ISSUE THREAD

**Issue by Braden-Preston** [HarvestHaven/TPOT-Toolbox#4]
Main issue is mentioned [here](https://github.com/storybooks/storybook/issues/6069): (**currently unresolved**)

**The Gist:**

Storybook tries to be smart and load up react-scripts in order to pull its webpack.config and .babelrc. By default react-scripts doesn't support decorators, so you have to override it, which Storybook allows, by putting your own webpack.config and .babelrc inside the .storybook root directory. 

This would work, however, Storybook 5 in its current configuration chooses to load up react-scripts AFTER loading up config overrides from the .storybook directory. Which means all the overrides you set get overridden anyways. This is only an issue with CRA bootstrapped projects.

One potential solution is to eject your whole CRA app, but then you are left with a difficult to understand set of configuration files. You can edit them to support decorators and then storybook, after realizing that there is no react-scripts anymore, will default to loading your ejected apps config and then properly, the overrides defined in .storybook. This is all swell.

But ejecting is not a practical option. That is why we use craco to provide overrides to react-scripts without ejecting. However, storybook doesn't let you use anything other than react-scripts and it doesn't allow you to change the default run command. Ie: I would like to change the run command to be yarn craco start, but by default, storybook only allows npm react-scripts start.

There is a possibility that a number of things will change in the future. Either CRA will support decorators, Storybook will allow custom run scripts or some genius on GitHub will figure out how to fix the issue without ejecting.

**What this means:**

Unfortunately, we have to sit on our hands for now, which is that we will have to write out ALL code without MobX decorator syntax IF you want to develop with storybook. If you don't want to do that, you can just use decorators with the default yarn start dev server, you just won't be able to have the interoperability of developing react components quicky using Storybook. 

One possible solution may be to revert back to Storybook 4 and MobX 4, but that isn't ideal since most add-ons will be upgrading to Storybook 5 anyways.

**Comment by Braden-Preston** [comment]
Possible workaround. If the application is bootstrapped with ```npx create-react-app tpot-toolbox -typescript```, it will add TS support and subsequently decorator support to the react-scripts config internally. Which means it may not need to be overridden by Storybook because it is already integrated. 

From: https://mobx.js.org/best/decorators.html

### Decorator syntax and Create React App (v2)
- Decorators are ***[```only supported out of the box when using TypeScript in create-react-app@^2.1.1.```](url)*** In older versions or when using vanilla JavaScript use either the decorate utility, eject, or the customize-cra package.

Whether Typescript is used or not, its configuration could be set up and then if it isn't used, remove it from the build using Craco and tree shaking, so there are no unused modules.

**Comment by Braden-Preston** [comment]
Fixed in `storybook-5-decorator-support` c846a56720ce7b67be4ce4630030001646612848 !

Here is my writeup on how to fix this easily.

### How to add custom babel configuration to Storybook 5.0.1 when bootstrapped with Create React App 2.0.

Storybook searches your project directory to figure out what you bootstrapped with (CRA, Next.js, Parcel, etc.) In the case of CRA it looks for  `react-scripts` in your package.json. `react-scripts` internally has a webpack build configuration that includes presets for linting, class constructors, babel plugins and more.

In order to add decorator syntax support to a project, you need the babel plugin `@babel/plugin-proposal-decorators` with "legacy" set to `true`. However by default  `react-scripts` (CRA), has this defaulted to be `false`. This means we need to modify the web pack configuration to add this babel plugin appropriately. However, you cannot change the webpack contents of `react-scripts`, as it is an externally managed repository. You could fork it, but then you have to maintain a complex configuration. CRA also gives you an option to eject your app, which will expand the contents of `react-scripts` into your directory, but again this has its own problems in maintenance and complexity.

It would be better to not have to eject, but still have the ability to override your webpack configuration. Fortunately, Storybook allows several override configs in the `.storybook` directory. See [Custom Webpack Config](https://storybook.js.org/docs/configurations/custom-webpack-config/) and [Custom Babel Config](https://storybook.js.org/docs/configurations/custom-babel-config/)

> **IMPORTANT** Keep in mind the way that Storybook loads up configuration in order. First it will look for configs in your root project directory, like your `.babelrc` and `.tsconfig`. Next it looks for bootstrap configurations, like `react-scripts`, and finally it looks inside its own `.storybook` directory for final overrides. This means your `.babelrc` and other root directory configs will be overrided by whatever is found in the bootstrap configurations, like `react-scripts`. So if you need to add similar overrides you will need to have the same `.babelrc` in BOTH your home directory `./.babelrc` AND `./.storybook/.babelrc`.

Given this load order, in order to override `react-scripts` webpack and babel configurations, put the overrides in the `.storybook` directory, just like this (You need both the `.babelrc` and `webpack.config.js` for this to work):

#### .babelrc
```javascript
{
  "plugins": [
    [ "@babel/plugin-proposal-decorators", { "legacy": true } ],
    [ "@babel/plugin-proposal-class-properties", { "loose": true } ]
  ],
  "presets": [
    "react-app"
  ]
}
```

#### webpack.config.js

```javascript
module.exports = async ({ config, mode }) => {

  // Decorator Syntax Support for MobX and Typescript
  config.module.rules.push({
    test: /(\.js|\.jsx|\.stories\.jsx)$/,
    exclude: /node_modules/,
    use: {
      loader: "babel-loader",
      options: {
        presets: [
          "@babel/preset-env",
          "@babel/preset-react"
        ],
        plugins: [
          ["@babel/plugin-proposal-decorators", { "legacy": true}],
          ["@babel/plugin-proposal-class-properties", { "loose": true}]
        ],
      },
    },
  });

  // Return the altered config
  return config;
}
```

Also, in your `package.json`, the following packages and versions:

#### package.json

``` json
{
  "dependencies": {
    "react-scripts": "2.1.8"
  },
  "devDependencies": {
    "@babel/plugin-proposal-class-properties": "7.3.4",
    "@babel/plugin-proposal-decorators": "7.3.0",
    "@storybook/react": "5.0.1"
  }
}
```

This guide should help integrate not just decorator syntax support into Storybook, but theoretically any webpack or babel plugin, transform, preset, etc. Hopefully, this helps.
